File: examples/python/03-cost-estimation-text/estimate_from_text.py
"""
Cost estimation from a text description of construction work.

Uses an LLM to decompose a description into work items, searches each
in Qdrant, and produces a structured cost estimate.
"""

import argparse
import json
import os
import sys
from pathlib import Path

try:
    from dotenv import load_dotenv
except ImportError:
    sys.exit("python-dotenv is required. Install with: pip install python-dotenv")

load_dotenv(Path(__file__).resolve().parents[2] / ".env")
load_dotenv()

try:
    from openai import OpenAI
except ImportError:
    sys.exit("openai is required. Install with: pip install openai")

try:
    from qdrant_client import QdrantClient
except ImportError:
    sys.exit("qdrant-client is required. Install with: pip install qdrant-client")

EMBEDDING_MODEL = "text-embedding-3-large"
DEFAULT_COLLECTION = "ddc_en_toronto"
QDRANT_URL = os.getenv("QDRANT_URL", "http://localhost:6333")

DECOMPOSE_PROMPT = """You are a construction cost estimator. Given a project description,
decompose it into individual work items that would appear in a Bill of Quantities.

For each work item, provide:
- description: short name of the work item (suitable for database search)
- quantity: estimated quantity (number)
- unit: measurement unit (m², m³, m, pcs, kg, etc.)

Return ONLY valid JSON as a list of objects. Example:
[
  {"description": "ceramic floor tiling", "quantity": 50, "unit": "m²"},
  {"description": "wall waterproofing membrane", "quantity": 60, "unit": "m²"}
]

Project description: {description}"""


def decompose_with_openai(description: str) -> list[dict]:
    """Use GPT-4o to decompose the description into work items."""
    client = OpenAI(api_key=os.getenv("OPENAI_API_KEY"))
    response = client.chat.completions.create(
        model="gpt-4o",
        messages=[{"role": "user", "content": DECOMPOSE_PROMPT.format(description=description)}],
        temperature=0.2,
    )
    return json.loads(response.choices[0].message.content.strip().strip("```json").strip("```"))


def decompose_with_anthropic(description: str) -> list[dict]:
    """Use Claude to decompose the description into work items."""
    try:
        import anthropic
    except ImportError:
        sys.exit("anthropic is required. Install with: pip install anthropic")

    client = anthropic.Anthropic(api_key=os.getenv("ANTHROPIC_API_KEY"))
    response = client.messages.create(
        model="claude-sonnet-4-20250514",
        max_tokens=2048,
        messages=[{"role": "user", "content": DECOMPOSE_PROMPT.format(description=description)}],
    )
    text = response.content[0].text.strip().strip("```json").strip("```")
    return json.loads(text)


def embed_text(text: str) -> list[float]:
    """Embed text using OpenAI."""
    client = OpenAI(api_key=os.getenv("OPENAI_API_KEY"))
    response = client.embeddings.create(model=EMBEDDING_MODEL, input=text)
    return response.data[0].embedding


def search_work_item(qdrant: QdrantClient, collection: str, description: str, top_k: int = 1) -> dict | None:
    """Search Qdrant for the best matching work item."""
    vector = embed_text(description)
    results = qdrant.search(collection_name=collection, query_vector=vector, limit=top_k)
    if not results:
        return None
    return results[0].payload


def main():
    parser = argparse.ArgumentParser(description="Estimate construction cost from text description.")
    parser.add_argument("description", type=str, help="Project description in natural language.")
    parser.add_argument("--provider", choices=["anthropic", "openai"], default=None,
                        help="LLM provider. Auto-detects from available API keys.")
    parser.add_argument("--collection", type=str, default=DEFAULT_COLLECTION, help="Qdrant collection.")
    parser.add_argument("--qdrant-url", type=str, default=QDRANT_URL, help="Qdrant URL.")
    args = parser.parse_args()

    # Determine LLM provider
    provider = args.provider
    if not provider:
        if os.getenv("ANTHROPIC_API_KEY"):
            provider = "anthropic"
        elif os.getenv("OPENAI_API_KEY"):
            provider = "openai"
        else:
            sys.exit("Set ANTHROPIC_API_KEY or OPENAI_API_KEY in .env.")

    if provider == "openai" and not os.getenv("OPENAI_API_KEY"):
        sys.exit("OPENAI_API_KEY required for OpenAI provider and embeddings.")
    if provider == "anthropic" and not os.getenv("ANTHROPIC_API_KEY"):
        sys.exit("ANTHROPIC_API_KEY required for Anthropic provider.")
    # Embeddings always need OpenAI
    if not os.getenv("OPENAI_API_KEY"):
        sys.exit("OPENAI_API_KEY is required for embeddings (text-embedding-3-large).")

    qdrant = QdrantClient(url=args.qdrant_url)

    # Step 1: Decompose
    print(f"Project: \"{args.description}\"")
    print(f"LLM provider: {provider}\n")
    print("Decomposing into work items...")

    if provider == "anthropic":
        work_items = decompose_with_anthropic(args.description)
    else:
        work_items = decompose_with_openai(args.description)

    print(f"Found {len(work_items)} work items.\n")

    # Step 2: Search and price each item
    estimate_lines = []
    total_material = 0.0
    total_labor = 0.0
    total_equipment = 0.0
    total_cost = 0.0

    print(f"{'#':<4} {'Description':<40} {'Qty':<8} {'Unit':<6} {'Unit Rate':>12} {'Amount':>14}")
    print("-" * 90)

    for i, item in enumerate(work_items, 1):
        desc = item["description"]
        qty = float(item.get("quantity", 1))
        unit = item.get("unit", "pcs")

        match = search_work_item(qdrant, args.collection, desc)

        if match:
            rate_name = match.get("rate_original_name", desc)
            unit_cost = float(match.get("total_cost_per_position", 0))
            material_cost = float(match.get("total_material_cost_per_position", 0))
            resource_cost = float(match.get("total_resource_cost_per_position", 0))
            equip_cost = float(match.get("total_value_machinery_equipment", 0))
            amount = unit_cost * qty
        else:
            rate_name = desc
            unit_cost = 0
            material_cost = 0
            resource_cost = 0
            equip_cost = 0
            amount = 0

        total_material += material_cost * qty
        total_labor += resource_cost * qty
        total_equipment += equip_cost * qty
        total_cost += amount

        print(f"{i:<4} {rate_name[:40]:<40} {qty:<8.1f} {unit:<6} {unit_cost:>12,.2f} {amount:>14,.2f}")

        estimate_lines.append({
            "item": i,
            "description": rate_name,
            "quantity": qty,
            "unit": unit,
            "unit_rate": unit_cost,
            "amount": amount,
        })

    # Summary
    print("-" * 90)
    print(f"{'':>60} {'TOTAL':>12} {total_cost:>14,.2f} EUR")
    print()
    print("=== Cost Breakdown ===")
    print(f"  Materials:  {total_material:>14,.2f} EUR")
    print(f"  Labor:      {total_labor:>14,.2f} EUR")
    print(f"  Equipment:  {total_equipment:>14,.2f} EUR")
    print(f"  Total:      {total_cost:>14,.2f} EUR")
    print()
    print("Note: Costs are based on DDC CWICR reference data. Actual costs may vary.")


if __name__ == "__main__":
    main()

File: 0_Workflow and Pipelines CWICR/javascript/04-cost-estimation-text/estimate_from_text.mjs
/**
 * 04 - Cost Estimation from Text Description
 *
 * Decomposes a free-text construction task into individual work items using
 * OpenAI, searches the DDC-CWICR Qdrant database for matching cost rates,
 * and produces a structured cost estimate.
 *
 * Usage:
 *   export OPENAI_API_KEY="sk-..."
 *   node estimate_from_text.mjs
 */

import { QdrantClient } from "@qdrant/js-client-rest";
import OpenAI from "openai";

// ---------------------------------------------------------------------------
// Configuration
// ---------------------------------------------------------------------------

const QDRANT_URL = process.env.QDRANT_URL || "http://localhost:6333";
const QDRANT_API_KEY = process.env.QDRANT_API_KEY || undefined;
const COLLECTION = "ddc_en_toronto";
const EMBEDDING_MODEL = "text-embedding-3-large";
const EMBEDDING_DIMS = 3072;
const TOP_K = 3;

const TASK_DESCRIPTION = `
Build a reinforced concrete foundation for a two-story residential house.
The foundation is 12 m x 10 m, 0.4 m thick, with steel rebar reinforcement.
Include formwork, concrete pouring, and waterproofing membrane.
`;

// ---------------------------------------------------------------------------
// Initialize clients
// ---------------------------------------------------------------------------

const openai = new OpenAI();

const qdrant = new QdrantClient({
  url: QDRANT_URL,
  ...(QDRANT_API_KEY && { apiKey: QDRANT_API_KEY }),
});

// ---------------------------------------------------------------------------
// Step 1 — Decompose the task into searchable work items
// ---------------------------------------------------------------------------

async function decomposeTask(description) {
  console.log("Step 1: Decomposing task into work items...\n");

  const response = await openai.chat.completions.create({
    model: "gpt-4o",
    temperature: 0,
    messages: [
      {
        role: "system",
        content: `You are a construction cost estimator. Given a construction task
description, decompose it into individual work items that can be searched in a
construction cost database. Return a JSON array of objects with:
- "item": short description of the work item (suitable for semantic search)
- "quantity": estimated quantity (number)
- "unit": unit of measurement (e.g., m2, m3, kg, pcs, lm)

Return ONLY valid JSON, no markdown fences.`,
      },
      {
        role: "user",
        content: description.trim(),
      },
    ],
  });

  const items = JSON.parse(response.choices[0].message.content);
  console.log(`  Identified ${items.length} work items:\n`);
  for (const item of items) {
    console.log(`    - ${item.item} (${item.quantity} ${item.unit})`);
  }
  console.log();
  return items;
}

// ---------------------------------------------------------------------------
// Step 2 — Search Qdrant for each work item
// ---------------------------------------------------------------------------

async function searchWorkItem(item) {
  const embResponse = await openai.embeddings.create({
    model: EMBEDDING_MODEL,
    input: item.item,
    dimensions: EMBEDDING_DIMS,
  });
  const vector = embResponse.data[0].embedding;

  const results = await qdrant.search(COLLECTION, {
    vector,
    limit: TOP_K,
    with_payload: true,
  });

  return results;
}

// ---------------------------------------------------------------------------
// Step 3 — Build the estimate
// ---------------------------------------------------------------------------

async function buildEstimate(workItems) {
  console.log("Step 2: Searching database for matching cost rates...\n");
  console.log("Step 3: Building cost estimate...\n");

  const estimate = [];
  let grandTotal = 0;

  for (const item of workItems) {
    const results = await searchWorkItem(item);

    // Use the best match (highest score)
    const best = results[0];
    if (!best) {
      estimate.push({
        workItem: item.item,
        quantity: item.quantity,
        unit: item.unit,
        match: "No match found",
        unitCost: 0,
        totalCost: 0,
        score: 0,
      });
      continue;
    }

    const payload = best.payload || {};
    const unitCost = Number(payload.total_cost_per_position) || 0;
    const materialCost = Number(payload.total_material_cost_per_position) || 0;
    const totalCost = unitCost * item.quantity;
    grandTotal += totalCost;

    estimate.push({
      workItem: item.item,
      quantity: item.quantity,
      unit: item.unit,
      matchedRate: payload.rate_final_name || payload.rate_original_name || "N/A",
      rateCode: payload.rate_code || "N/A",
      rateUnit: payload.rate_unit || "N/A",
      unitCost,
      materialCost,
      totalCost,
      score: best.score,
    });
  }

  return { lineItems: estimate, grandTotal };
}

// ---------------------------------------------------------------------------
// Display
// ---------------------------------------------------------------------------

function displayEstimate(result) {
  console.log("=".repeat(80));
  console.log("  CONSTRUCTION COST ESTIMATE");
  console.log("=".repeat(80));
  console.log();

  for (const [i, line] of result.lineItems.entries()) {
    console.log(`  ${i + 1}. ${line.workItem}`);
    console.log(`     Matched Rate : ${line.matchedRate || "N/A"} [${line.rateCode}]`);
    console.log(`     Match Score  : ${(line.score || 0).toFixed(4)}`);
    console.log(`     Quantity     : ${line.quantity} ${line.unit}`);
    console.log(`     Unit Cost    : EUR ${line.unitCost.toFixed(2)} / ${line.rateUnit}`);
    console.log(`     Line Total   : EUR ${line.totalCost.toFixed(2)}`);
    console.log();
  }

  console.log("-".repeat(80));
  console.log(`  GRAND TOTAL: EUR ${result.grandTotal.toFixed(2)}`);
  console.log("-".repeat(80));
  console.log();
  console.log(
    "  Note: This is an indicative estimate based on the closest matching"
  );
  console.log(
    "  rates in the DDC-CWICR database. Actual costs may vary."
  );
  console.log();
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

async function main() {
  console.log("\nTask Description:");
  console.log(`  ${TASK_DESCRIPTION.trim()}\n`);

  const workItems = await decomposeTask(TASK_DESCRIPTION);
  const estimate = await buildEstimate(workItems);
  displayEstimate(estimate);
}

main().catch((err) => {
  console.error("Error:", err);
  process.exit(1);
});
